File: mobile/src/main/java/org/openhab/habdroid/ui/widget/UrlInputPreference.java
package org.openhab.habdroid.ui.widget;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;
import android.preference.EditTextPreference;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.StringRes;

import org.openhab.habdroid.R;

import java.net.MalformedURLException;
import java.net.URL;

public class UrlInputPreference extends EditTextPreference implements TextWatcher {
    private EditText mEditor;
    private boolean mUrlIsValid;

    public UrlInputPreference(Context context) {
        super(context);
    }

    public UrlInputPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public UrlInputPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onAddEditTextToDialogView(View dialogView, EditText editText) {
        super.onAddEditTextToDialogView(dialogView, editText);
        mEditor = editText;
        editText.addTextChangedListener(this);
    }

    @Override
    protected void showDialog(Bundle state) {
        super.showDialog(state);
        updateOkButtonState();
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int start, int before, int count) {
    }

    @Override
    public void onTextChanged(CharSequence charSequence, int start, int before, int count) {
    }

    @Override
    public void afterTextChanged(Editable editable) {
        boolean portSeemsInvalid = false;
        if (TextUtils.isEmpty(editable)) {
            mUrlIsValid = true;
        } else {
            String value = editable.toString();
            if (value.contains("\n") || value.contains(" ")) {
                mUrlIsValid = false;
            } else {
                try {
                    URL url = new URL(value);
                    mUrlIsValid = true;
                    if (url.getProtocol().equals("http")) {
                        portSeemsInvalid = url.getPort() == 443 || url.getPort() == 8443;
                    }
                    if (url.getProtocol().equals("https")) {
                        portSeemsInvalid = url.getPort() == 80 || url.getPort() == 8080;
                    }
                } catch (MalformedURLException e) {
                    mUrlIsValid = false;
                }
            }
        }
        @StringRes int error = 0;
        if (!mUrlIsValid) {
            error = R.string.error_invalid_url;
        } else if (portSeemsInvalid) {
            error = R.string.error_port_seems_invalid;
        }
        mEditor.setError(error == 0 ? null : mEditor.getResources().getString(error));
        updateOkButtonState();
    }

    private void updateOkButtonState() {
        if (getDialog() instanceof AlertDialog) {
            AlertDialog dialog = (AlertDialog) getDialog();
            Button okButton = dialog.getButton(AlertDialog.BUTTON_POSITIVE);
            if (okButton != null) {
                okButton.setEnabled(mUrlIsValid);
            }
        }
    }
}

File: mobile/src/main/java/org/openhab/habdroid/ui/widget/UrlInputPreference.kt
package org.openhab.habdroid.ui.widget

import android.app.AlertDialog
import android.content.Context
import android.os.Bundle
import android.preference.EditTextPreference
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.View
import android.widget.EditText
import org.openhab.habdroid.R
import java.net.MalformedURLException
import java.net.URL

class UrlInputPreference constructor(context: Context, attrs: AttributeSet) :
    EditTextPreference(context, attrs), TextWatcher {
    private lateinit var editor: EditText
    private var urlIsValid: Boolean = false

    override fun onAddEditTextToDialogView(dialogView: View, editText: EditText) {
        super.onAddEditTextToDialogView(dialogView, editText)
        editor = editText
        editText.addTextChangedListener(this)
    }

    override fun showDialog(state: Bundle?) {
        super.showDialog(state)
        updateOkButtonState()
    }

    override fun beforeTextChanged(charSequence: CharSequence, start: Int, before: Int, count: Int) {
        // no-op
    }

    override fun onTextChanged(charSequence: CharSequence, start: Int, before: Int, count: Int) {
        // no-op
    }

    override fun afterTextChanged(editable: Editable) {
        var portSeemsInvalid = false
        if (editable.isEmpty()) {
            urlIsValid = true
        } else {
            val value = editable.toString()
            if (value.contains("\n") || value.contains(" ")) {
                urlIsValid = false
            } else {
                try {
                    val url = URL(value)
                    urlIsValid = true
                    when (url.protocol) {
                        "http" -> portSeemsInvalid = url.port == 443 || url.port == 8443
                        "https" -> portSeemsInvalid = url.port == 80 || url.port == 8080
                    }
                } catch (e: MalformedURLException) {
                    urlIsValid = false
                }
            }
        }
        val res = editor.resources
        editor.error = when {
            !urlIsValid -> res.getString(R.string.error_invalid_url)
            portSeemsInvalid -> res.getString(R.string.error_port_seems_invalid)
            else -> null
        }
        updateOkButtonState()
    }

    private fun updateOkButtonState() {
        val dialog = this.dialog
        if (dialog is AlertDialog) {
            dialog.getButton(AlertDialog.BUTTON_POSITIVE)?.isEnabled = urlIsValid
        }
    }
}
